posix/select.c

```diff
@@ -0,0 +1,237 @@
+        {
+            OE_FD_SET(pevent[i].data.fd, set);
+        }
+    }
+}
+
+int oe_select(
+    int nfds,
+    oe_fd_set* readfds,
+    oe_fd_set* writefds,
+    oe_fd_set* exceptfds,
+    struct oe_timeval* timeout)
+{
+    int fd_list[OE_FD_SETSIZE] = {0};  // <nfds> members
+    int fd_flags[OE_FD_SETSIZE] = {0}; // indexed by fd
+    int i = 0;
+    int epoll_fd = -1;
+    int ret = -1;
+    int ret_fds = -1;
+    struct oe_epoll_event rtn_ev[OE_FD_SETSIZE] = {{0}};
+    int timeout_ms = -1;
+
+    if (timeout)
+    {
+        timeout_ms = (int)timeout->tv_sec * 1000;
+        timeout_ms += (int)(timeout->tv_usec / 1000);
+    }
+
+    {
+        const size_t num_bytes = sizeof(uint32_t) * (size_t)(nfds + 1);
+
+        if (oe_memset_s(fd_list, num_bytes, 0xff, num_bytes) != OE_OK)
+            OE_RAISE_ERRNO(OE_EINVAL);
+    }
+
+    if (readfds)
+    {
+        _set_to_fd_list(
+            readfds,
+            (OE_EPOLLIN | OE_EPOLLRDNORM | OE_EPOLLRDBAND),
+            nfds,
+            fd_list,
+            fd_flags);
+    }
+
+    if (writefds)
+    {
+        _set_to_fd_list(
+            writefds,
+            (OE_EPOLLOUT | OE_EPOLLWRNORM | OE_EPOLLWRBAND),
+            nfds,
+            fd_list,
+            fd_flags);
+    }
+
+    if (exceptfds)
+    {
+        _set_to_fd_list(
+            exceptfds,
+            (OE_EPOLLERR | OE_EPOLLHUP | OE_EPOLLRDHUP | OE_EPOLLWAKEUP),
+            nfds,
+            fd_list,
+            fd_flags);
+    }
+
+    epoll_fd = oe_epoll_create1(0);
+    if (epoll_fd < 0)
+        OE_RAISE_ERRNO(oe_errno);
+
+    for (i = 0; i < nfds; i++)
+    {
+        struct oe_epoll_event ev = {
+            .data.fd = fd_list[i],
+            .events = (uint32_t)fd_flags[fd_list[i]],
+        };
+
+        ret = oe_epoll_ctl(epoll_fd, OE_EPOLL_CTL_ADD, fd_list[i], &ev);
+        if (ret < 0)
+        {
+            OE_RAISE_ERRNO_MSG(
+                OE_EINVAL,
+                "epoll_fd=%d fd_list[%d]=%d",
+                epoll_fd,
+                i,
+                fd_list[i]);
+        }
+    }
+
+    ret_fds = oe_epoll_wait(epoll_fd, rtn_ev, OE_FD_SETSIZE, timeout_ms);
+    if (ret_fds < 0)
+    {
+        goto done;
+    }
+
+    if (readfds)
+        OE_FD_ZERO(readfds);
+
+    if (writefds)
+        OE_FD_ZERO(writefds);
+
+    if (exceptfds)
+        OE_FD_ZERO(exceptfds);
+
+    if (readfds)
+    {
+        _ev_list_to_set(
+            ret_fds,
+            rtn_ev,
+            (OE_EPOLLIN | OE_EPOLLRDNORM | OE_EPOLLRDBAND),
+            readfds);
+    }
+
+    if (writefds)
+    {
+        _ev_list_to_set(
+            ret_fds,
+            rtn_ev,
+            (OE_EPOLLOUT | OE_EPOLLWRNORM | OE_EPOLLWRBAND),
+            writefds);
+    }
+
+    if (exceptfds)
+    {
+        _ev_list_to_set(
+            ret_fds,
+            rtn_ev,
+            (OE_EPOLLERR | OE_EPOLLHUP | OE_EPOLLRDHUP | OE_EPOLLWAKEUP),
+            exceptfds);
+    }
+done:
+    oe_close(epoll_fd);
```

Comment:
Is a check for epoll_fd >=0 required before calling oe_close?

Reply:
Thanks. I added the check. Bruce might have decided to let **oe_close()** detect the error.